### PR TITLE
Allow to change SSH password

### PR DESCRIPTION
**Pull request by AaronDewes**
This is just the host part, but it's a first step.

Fixes: Twitter users who will tweet about this issue until a fix is released, ﻿ #381.

### CODE REVIEW

**Comment by AaronDewes** [comment]
@lukechilds Do you a know a way to pass the password encrypted? The password will be on the filesystem for a few seconds with this, and I'm not sure if you're comfortable with that, especially if the SSH password should be the same as the user password, but if this amount of time is after the login, an attacker with shell access could already modify lnd from CLI because it's unlocked

**Comment by lukechilds** [comment]
There's no need to pass the password encrypted, if we do then we need to pass the decryption key along too or use a hardcoded one. Either way, if an attacker has access to the environment to view the encrypted password it would be trivial to decrypt.

**Comment by louneskmt** [comment]
I don't think this is a good way to communicate the new password to the system, as it may contain special characters, and file names cannot contain some of them.

**Comment by AaronDewes** [comment]
>ou're right, I'll put it into the file

File: events/triggers/passwd [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+signal="${1}"
+
+# Try to load Umbrel OS version info to detect Umbrel OS
+[[ -f "/etc/default/umbrel" ]] && source "/etc/default/umbrel"
+
+
+# Only change password on Umbrel OS
+if [[ ! -z "${UMBREL_OS:-}" ]]; then
+    # Script runs as root, we need to set the username
+    echo ${1} | passwd umbrel --stdin
```
Comment (Kixunil) [inline events/triggers/passwd:12]: Did you mean to write `cat "$signal" | passwd umbrel --stdin`?

I don't know much about Umbrel but it seems to me you're storing this inside the file not in the file name?

File: events/triggers/passwd [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+signal="${1}"
```
Comment (Kixunil) [inline events/triggers/passwd:3]: Passing password on command line is not great but not too bad given current assumptions. Make sure to add this to your TODO list of things to fix when implementing isolation.

File: karen [hunk 3]
```diff
@@ -63,6 +63,14 @@ fswatch -0 --event=PlatformSpecific $signal_dir | while read -d "" event; do
     args="${signal#$app_prefix}"
   fi
 
+  pw_prefix="passwd-"
+  if [[ "$signal" == "$pw_prefix"* ]]; then
+    trigger="$trigger_dir/passwd"
+    args="${signal#$pw_prefix}"
+    # Delete signal to remove pw from file system
+    rm "$signal_dir/$signal"
```
Comment (Kixunil) [inline karen:71]: If you're worried about it being stored on disk then I recommend using `/dev/shm/` for this, which is stored in RAM and will be definitely wiped at worst after reboot. If you need some file in directory, then symlink will solve this.

**Review by Kixunil — COMMENTED** [review]
Doesn't seem entirely correct to me. Do you have your events/triggers documented somewhere so that I could help by reviewing it more properly?

File: events/triggers/passwd (hunk 1 above)
Comment (AaronDewes) [reply events/triggers/passwd:12]: I already said that I will change it.

File: karen (hunk 3 above)
Comment (AaronDewes) [reply karen:71]: I think the current solution is okay.

Comment (Kixunil) [reply karen:71]: Yes, not a big deal at this stage of development.

**Comment by lukechilds** [comment]
Thanks for the review @Kixunil.

>Doesn't seem entirely correct to me. Do you have your events/triggers documented somewhere so that I could help by reviewing it more properly?

We don't have it documented, it's basically just when `events/signals/foo` is touched `events/trigger/foo` is executed. There was some special handling for apps. This was just to quickly allow Docker containers to run certain jobs on the host. A proper RPC server is planned long term.

@AaronDewes the way we usually pass data around is in status files. So we should probably have some sort of `status/password-updated` file which by default is `false`. We then write the password to that file, touch `events/signals/change-password` which causes `events/triggers/change-password` to be executed which reads `status/password-updated` and updates the Umbrel user password with the value, then sets the value of `status/password-updated` to `true`.

This actually introduces a new vulnerability where if an attacker can somehow execute code on the host as a non-umbrel non-root user they can escalate their privileges to umbrel (and then root via sudo) by writing to the signal/status files to set the umbrel password to something they know. This is not an issue right now since it's still better than the current setup, docker provides (some) isolation from compromised services being able to execute on the host so this attack would not at all be easy to pull off. This can be resolved properly with the new RPC server by making sure it only responds to requests with a valid token and tokens are only readable by select users/containers with the required permissions.

Long term this will all be handled at setup and future password changes will be handled more cleanly by an RPC server running on the host without need for writing persistent files. The RPC server will only responds to requests with a valid token and tokens will only be readable by select users/containers with the required permissions.

**Comment by AaronDewes** [comment]
I hope this is better now (the host part, will implement the manager part later)

**Review by lukechilds — APPROVED** [review]

**Comment by lukechilds** [comment]
Thanks @AaronDewes!

**Comment by AaronDewes** [comment]
I forgot about htis PR, will implement the manatger part today

**Pull request merged**